File: src/test/java/libsys/UserTest.java
package libsys;

import junit.framework.Test;
import junit.framework.TestCase;
import junit.framework.TestSuite;

import java.util.ArrayList;

/**
 * Test User related operations
 */
public class UserTest extends TestCase
{
    private User user1;
    private User user2;

    /**
     * Test User related operations
     * @param testName name of the test case
     */
    public UserTest(String testName)
    {
        super(testName);
    }

    /**
     * @return suite of the tests being tested
     */
    public static Test suite()
    {
        return new TestSuite(UserTest.class);
    }

    /**
     * Run all tests
     */
    public void testApp()
    {
        valueTest();
        setterTest();
        bookInteractionTest();
    }

    /**
     * Set up before testing
     * @throws Exception Exception
     */
    public void setUp() throws Exception
    {
        super.setUp();
        user1 = new User("User1", 0, 10, new ArrayList<Integer>());
        ArrayList<Integer> user2List = new ArrayList<Integer>();
        user2List.add(0);
        user2 = new User("User2", 2, 8, user2List);
    }

    /**
     * Test if the values are accurate
     */
    private void valueTest()
    {
        assertEquals("Name is \"User1\"", user1.getName(), "User1");
        assertEquals("id is 0", user1.getId(), 0);
        assertEquals("Limit is 10", user1.getLimit(), 10);
        int[] bookIds = user1.bookStatus();
        assertEquals("Book list is empty", bookIds.length, 0);

        assertEquals("Name is \"User2\"", user2.getName(), "User2");
        assertEquals("id is 2", user2.getId(), 2);
        assertEquals("Limit is 8", user2.getLimit(), 8);
        bookIds = user2.bookStatus();
        assertEquals("Book list is empty", bookIds.length, 1);
        assertEquals("Book id is 0", bookIds[0], 0);
    }

    /**
     * Test the setter functions
     */
    private void setterTest()
    {
        String newName = "New User 1";
        user1.setName(newName);
        assertEquals("New name for User1 is \"New User 1\"", user1.getName(), newName);
        user1.setLimit(1);
        assertEquals("New limit for User1 is 8", user1.getLimit(), 1);

        newName = "New User 2";
        user2.setName(newName);
        assertEquals("New name for User2 is \"New User 2\"", user2.getName(), newName);
        user2.setLimit(3);
        assertEquals("New limit for User2 is 3", user2.getLimit(), 3);
    }

    /**
     * Test all functions related to Book
     */
    private void bookInteractionTest()
    {
        assertTrue(user1.borrowNewBook(3));
        int[] bookIds = user1.bookStatus();
        assertEquals("User1 is now borrowing Book3", bookIds[0], 3);
        assertFalse("User1 reached the limit", user1.status());
        assertFalse("User1 failed to return Book2 that was not borrowed", user1.returnBook(2));
        assertTrue("User1 return Book3", user1.returnBook(3));
        assertTrue("User1 can now borrow book", user1.status());
        assertEquals("User1 is not borrowing anything", user1.bookStatus().length, 0);
    }
}

File: src/test/kotlin/libsys/UserTest.kt
package libsys

import org.junit.*
import java.util.ArrayList

/**
 * Test User related operations
 */
class UserTest
{
    private var user1: User? = null
    private var user2: User? = null

    @Before fun prepareTest()
    {
        user1 = User("User1", 0, 10, ArrayList<Int>())
        val user2List = ArrayList<Int>()
        user2List.add(0)
        user2 = User("User2", 2, 8, user2List)
    }

    /**
     * Test User related operations
     */
    @Test fun userTest()
    {
        Assert.assertEquals("Name is \"User1\"", user1!!.name, "User1")
        Assert.assertEquals("id is 0", user1!!.id, 0)
        Assert.assertEquals("Limit is 10", user1!!.limit, 10)
        var bookIds: IntArray = user1!!.bookStatus()
        Assert.assertEquals("Book list is empty", bookIds.size, 0)

        Assert.assertEquals("Name is \"User2\"", user2!!.name, "User2")
        Assert.assertEquals("id is 2", user2!!.id, 2)
        Assert.assertEquals("Limit is 8", user2!!.limit, 8)
        bookIds = user2!!.bookStatus()
        Assert.assertEquals("Book list is empty", bookIds.size, 1)
        Assert.assertEquals("Book id is 0", bookIds[0], 0)

        var newName = "New User 1"
        user1!!.name = newName
        Assert.assertEquals("New name for User1 is \"New User 1\"", user1!!.name, newName)
        user1!!.limit = 1
        Assert.assertEquals("New limit for User1 is 8", user1!!.limit, 1)

        newName = "New User 2"
        user2!!.name = newName
        Assert.assertEquals("New name for User2 is \"New User 2\"", user2!!.name, newName)
        user2!!.limit = 3
        Assert.assertEquals("New limit for User2 is 3", user2!!.limit, 3)

        Assert.assertTrue(user1!!.borrowNewBook(3))
        bookIds = user1!!.bookStatus()
        Assert.assertEquals("User1 is now borrowing Book3", bookIds[0], 3)
        Assert.assertFalse("User1 reached the limit", user1!!.status())
        Assert.assertFalse("User1 fail to borrow another book", user1!!.borrowNewBook(5))
        Assert.assertFalse("User1 failed to return Book2 that was not borrowed", user1!!.returnBook(2))
        Assert.assertTrue("User1 return Book3", user1!!.returnBook(3))
        Assert.assertTrue("User1 can now borrow book", user1!!.status())
        Assert.assertEquals("User1 is not borrowing anything", user1!!.bookStatus().size, 0)
    }
}